src/validators/SessionKey/SessionKeyManager.sol

```diff
@@ -0,0 +1,147 @@
+        return userSessions[smartAccount];
+    }
+
+    /**
+     * @dev sets the merkle root of a tree containing session keys for msg.sender
+     * should be called by Smart Account
+     * @param _merkleRoot Merkle Root of a tree that contains session keys with their permissions and parameters
+     */
+    function setMerkleRoot(bytes32 _merkleRoot) external {
+        userSessions[msg.sender].merkleRoot = _merkleRoot;
+    }
+
+    function _sessionMerkelLeaf(
+        uint256 validUntil,
+        uint256 validAfter,
+        address sessionValidationModule,
+        bytes memory sessionKeyData
+    )
+        public
+        pure
+        returns (bytes32)
+    {
+        return keccak256(
+            abi.encodePacked(validUntil, validAfter, sessionValidationModule, sessionKeyData)
+        );
+    }
+
+    /**
+     * @dev validates userOperation
+     * @param userOp User Operation to be validated.
+     * @param userOpHash Hash of the User Operation to be validated.
+     * @return sigValidationResult 0 if signature is valid, SIG_VALIDATION_FAILED otherwise.
+     */
+    function validateUserOp(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    )
+        external
+        view
+        virtual
+        override
+        returns (uint256)
+    {
+        SessionStorage storage sessionKeyStorage = _getSessionData(userOp.sender);
+
+        SessionKeyParams memory sessionKeyParams =
+            abi.decode(userOp.decodeSignature(), (SessionKeyParams));
+
+        bytes32 leaf = _sessionMerkelLeaf({
+            validUntil: sessionKeyParams.validUntil,
+            validAfter: sessionKeyParams.validAfter,
+            sessionValidationModule: sessionKeyParams.sessionValidationModule,
+            sessionKeyData: sessionKeyParams.sessionKeyData
+        });
+        if (!MerkleProof.verify(sessionKeyParams.merkleProof, sessionKeyStorage.merkleRoot, leaf)) {
+            revert("SessionNotApproved");
+        }
+        //_packValidationData expects true if sig validation has failed, false otherwise
+        bool validSig = ISessionKeyValidationModule(sessionKeyParams.sessionValidationModule)
+            .validateSessionUserOp(
+            userOp,
+            userOpHash,
+            sessionKeyParams.sessionKeyData,
+            sessionKeyParams.sessionKeySignature
+        );
+
+        if (validSig) return 0;
+        else return 1;
+    }
+
+    /**
+     * @dev isValidSignature according to BaseAuthorizationModule
+     * @param _dataHash Hash of the data to be validated.
+     * @param _signature Signature over the the _dataHash.
+     * @return always returns 0xffffffff as signing messages is not supported by SessionKeys
+     */
+    function isValidSignature(
+        bytes32 _dataHash,
+        bytes memory _signature
+    )
+        public
+        view
+        override
+        returns (bytes4)
+    {
+        return 0xffffffff; // do not support it here
```

Comment:
Should we route this to the session key module?

Reply:
good call!